File: services/api/app/schemas/couple_retrospect.py
"""Schemas: Rueckblick ueber Zeit im Paarraum.

Die Statistik wird bei jedem Aufruf frisch gerechnet, Echos Text nur einmal erzeugt und
dann gelesen. Beides gehoert beiden Personen.
Siehe services/couple_retrospect_service.py.
"""
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


class CoupleMoodShare(BaseModel):
    mood: str
    anzahl: int


class CoupleRetrospectStats(BaseModel):
    period_start: date
    period_end: date
    days: int
    #: Durchschnitt BEIDER Barometer - nie die Tageskurve der anderen Person.
    barometer_avg: float | None = None
    barometer_avg_before: float | None = None
    barometer_delta: float | None = None
    moods: list[CoupleMoodShare]
    checkin_weeks: int
    sessions_started: int
    sessions_closed: int
    topics_opened: int
    topics_resolved: int
    agreements_made: int
    agreements_kept: int
    agreements_dropped: int
    appreciations: int
    #: Abgeschlossene Runden Ehrliches Mitteilen. Kein Ergebnis – eine Tatsache
    #: über die Praxis, wie die Check-in-Wochen daneben.
    honest_rounds: int = 0
    #: Reicht die Datenlage fuer einen sinnvollen Text?
    has_substance: bool


class CoupleRetrospective(BaseModel):
    id: UUID
    created_by: UUID
    period_start: date
    period_end: date
    body: str
    created_at: datetime


class CoupleRetrospectView(BaseModel):
    stats: CoupleRetrospectStats
    retrospectives: list[CoupleRetrospective]


class CoupleRetrospectCreate(BaseModel):
    days: int = Field(30, ge=14, le=365)

File: apps/web/src/api/coupleRetrospect.ts
import { apiClient } from './client'

/**
 * Rückblick über Zeit – was sich verändert hat, nicht nur dass ihr arbeitet.
 *
 * Die Zahlen werden bei jedem Aufruf frisch gerechnet; gespeichert wird nur Echos Text.
 * `barometer_avg` ist der Durchschnitt BEIDER – die Tageskurve der anderen Person gibt der
 * Server bewusst nie heraus.
 */
export interface CoupleRetrospectStats {
  period_start: string
  period_end: string
  days: number
  barometer_avg: number | null
  barometer_avg_before: number | null
  barometer_delta: number | null
  moods: { mood: string; anzahl: number }[]
  /** Abgeschlossene Runden Ehrliches Mitteilen. Kein Ergebnis – eine Tatsache
   *  über die Praxis, wie die Check-in-Wochen daneben. */
  honest_rounds: number
  checkin_weeks: number
  sessions_started: number
  sessions_closed: number
  topics_opened: number
  topics_resolved: number
  agreements_made: number
  agreements_kept: number
  agreements_dropped: number
  appreciations: number
  /** Reicht die Datenlage für einen sinnvollen Text? */
  has_substance: boolean
}

export interface CoupleRetrospective {
  id: string
  created_by: string
  period_start: string
  period_end: string
  body: string
  created_at: string
}

export interface CoupleRetrospectView {
  stats: CoupleRetrospectStats
  retrospectives: CoupleRetrospective[]
}

export const coupleRetrospectApi = {
  get: (coupleId: string, days = 30) =>
    apiClient.get<CoupleRetrospectView>(`/couple/links/${coupleId}/rueckblick`, {
      params: { days },
    }).then(r => r.data),

  write: (coupleId: string, days = 30) =>
    apiClient.post<CoupleRetrospective>(`/couple/links/${coupleId}/rueckblick`, { days })
      .then(r => r.data),

  remove: (retroId: string) =>
    apiClient.delete(`/couple/rueckblick/${retroId}`),
}
